Sources/Conduit/Networking/Data/DataDownloader.swift

```diff
@@ -0,0 +1,141 @@
+
+/// Represents an error that occurred within an DataDownloader
+/// - invalidRequest: An invalid request was supplied, most likely with an empty URL
+public enum DataDownloaderError: Error {
+    case invalidRequest
+}
+
+public protocol DataDownloaderType {
+    func downloadData(for request: URLRequest, completion: @escaping (DataDownloader.Response) -> Void) -> SessionTaskProxyType?
+}
+
+/// Utilizes Conduit to download and safely cache/retrieve
+/// data across multiple threads
+public final class DataDownloader: DataDownloaderType {
+
+    /// Represents a network or cached data response
+    public struct Response {
+        /// The resulting data
+        public let data: NSData?
+        /// The error that occurred from transport or cache retrieval
+        public let error: Error?
+        /// The URL response, if a download occurred
+        public let urlResponse: HTTPURLResponse?
+        /// Signifies if the data was retrieved directly from the cache
+        public let isFromCache: Bool
+
+        public init(data: NSData?, error: Error?, urlResponse: HTTPURLResponse?, isFromCache: Bool) {
+            self.data = data
+            self.error = error
+            self.urlResponse = urlResponse
+            self.isFromCache = isFromCache
+        }
+    }
+
+    /// A closure that fires upon data fetch success/failure
+    public typealias CompletionHandler = (Response) -> Void
+
+    private var cache: URLDataCache
+    private let sessionClient: URLSessionClientType
+    private var sessionProxyMap: [String: SessionTaskProxyType] = [:]
+    private var completionHandlerMap: [String: [CompletionHandler]] = [:]
+    private let completionQueue: OperationQueue?
+    private let serialQueue = DispatchQueue(
+        label: "com.mindbodyonline.Conduit.DataDownloader-\(UUID().uuidString)"
+    )
+
+    /// Initializes a new DataDownloader
+    /// - Parameters:
+    ///   - cache: The data cache in which to store downloaded data
+    ///   - sessionClient: The URLSessionClient to be used to download data
+    ///   - completionQueue: An optional operation queue for completion callback
+    public init(cache: URLDataCache,
+                sessionClient: URLSessionClientType = URLSessionClient(),
+                completionQueue: OperationQueue? = nil) {
+        self.cache = cache
+        self.sessionClient = sessionClient
+        self.completionQueue = completionQueue
+    }
+
+    /// Downloads data or retrieves it from the cache if previously downloaded.
+    /// - Parameters:
+    ///     - request: The request for the data
+    /// - Returns: A concrete SessionTaskProxyType
+    @discardableResult
+    public func downloadData(for request: URLRequest, completion: @escaping CompletionHandler) -> SessionTaskProxyType? {
+        var proxy: SessionTaskProxyType?
+        let completionQueue = self.completionQueue ?? .current ?? .main
+
+        serialQueue.sync { [weak self] in
+            guard let `self` = self else {
```

Comment:
If we unwrap `self` directly do we still need `[weak self]` defined above?

Reply:
The `[weak self]` is still needed, the guard just makes sure it's not nil before proceeding. Might be better to change name to `strongSelf` to clarify that rest of code will only run when `self` still exists

`guard let strongSelf = self else {`